src/async/ExtensiblePromise.ts

```diff
@@ -137,8 +139,8 @@ export default class ExtensiblePromise<T> {
 	 *
 	 * @returns {ExtensiblePromise}
 	 */
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => void): this;
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => (U | Thenable<U>)): this {
+	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => void): ExtensiblePromise<U>;
+	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => (U | Thenable<U>)): ExtensiblePromise<U> {
```

Comment:
I think there might be a problem with the types here as well. Consider the following example:

```typescript
import ExtensiblePromise from 'dojo-core/async/ExtensiblePromise';

const promise = new ExtensiblePromise((resolve, reject) => {
	reject(true);
});

promise.then(value => {
	// return nothing
}, () => return 'some string') // catch returns a string
.then(value => {
	console.log('length', value.length);
	// Property 'length' doesn't exist on type 'never';
});
```

Comparing this to how ES2015 Promises work, I wouldn't expect the `Property 'length' doesn't exist on type 'never';` to be there, as it's not guaranteed that the type is `never`. I haven't thoroughly tested this, but comparing our definition to the ES2015 Promise definition, this might be a solution:

```typescript
then(onFulfilled?: ((value: T) => T | Thenable<T>) | void, onRejected?: ((reason: Error) => T | Thenable<T>) | void): ExtensiblePromise<T>;
then<U>(onFulfilled: ((value: T) => Thenable<T>) | void, onRejected: (reason: Error) => U | Thenable<U>): ExtensiblePromise<T | U>;
then<U>(onFulfilled: (value: T) => U | Thenable<U>, onRejected?: ((reason: Error) => U | Thenable<U>) | void): ExtensiblePromise<U>;
then<U, V>(onFulfilled: (value: T) => U | Thenable<U>, onRejected: (reason: Error) => V | Thenable<V>): ExtensiblePromise<U | V> {
	const e: Executor<U | V> = (resolve, reject) => {
		function handler(rejected: boolean, valueOrError: T | U | Error) {
			const callback: ((value: T | U | V | Error) => (U | Thenable<U> | V | Thenable<V> | void)) | undefined = rejected ? onRejected : onFulfilled;

			if (typeof callback === 'function') {
				try {
					resolve(<U> callback(<T> valueOrError));
				}
				catch (error) {
					reject(error);
				}
			}
			else if (rejected) {
				reject(valueOrError);
			}
			else {
				resolve(<U> valueOrError);
			}
		}

		this._promise.then(handler.bind(null, false), handler.bind(null, true));
	};

	return new (<{ new(executor: Executor<U | V>): any }> this.constructor)(e);
}
```